Apply CA1835 - Prefer memory overloads for Stream.ReadAsync/WriteAsync

carlossanlop opened the pull request:
The CA1835 Roslyn analyzer/fixer has been merged to dotnet/roslyn-analyzers and now the rule needs to be applied in dotnet/runtime.

> Prefer Memory<T> overloads for Stream.Read/WriteAsync https://github.com/dotnet/runtime/issues/33790

stephentoub commented on src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextReaderImplAsync.cs:
```diff
@@ -975,7 +975,7 @@ private async Task InitStreamInputAsync(Uri baseUri, string baseUriStr, Stream s
             while (_ps.bytesUsed < 4 && _ps.bytes.Length - _ps.bytesUsed > 0)
             {
-                int read = await stream.ReadAsync(_ps.bytes, _ps.bytesUsed, _ps.bytes.Length - _ps.bytesUsed).ConfigureAwait(false);
+                int read = await stream.ReadAsync(_ps.bytes.AsMemory(_ps.bytesUsed, _ps.bytes.Length - _ps.bytesUsed)).ConfigureAwait(false);
```
We should ideally update the fixer to detect this. Namely if the count is array.Length - offset, use the single-parameter AsMemory instead.

carlossanlop replied:
I'll revert the two changes pointed out by @GrabYourPitchforks , update the fixer and then come back to runtime to make sure the fixer detects them for me. Sounds good?

carlossanlop replied:
Thinking about this more, I'll just apply it now, and when it comes time to test it in runtime, I'll temporarily revert this change locally to see if it would've worked, but that way I don't have to submit another PR.

carlossanlop replied:
@stephentoub I think we should also have a _separate_ Roslyn analyzer that detects when `buffer.AsMemory(offset, buffer.Length - offset)` is being used, and suggest using `buffer.AsMemory(offset)`, unless such analyzer already exists. What do you think? I can open the issue if you're ok with that.

carlossanlop replied:
I opened [an issue ](https://github.com/dotnet/runtime/issues/35981) suggesting the creation of an analyzer/fixer for `AsMemory`. Let me know what you think.

stephentoub replied:
I think it's reasonable in addition.  We should ensure this fixer generates the best code we can muster, but then being able to address other inefficiencies is useful, too.

carlossanlop replied:
Awesome. I also opened [this issue](https://github.com/dotnet/roslyn-analyzers/issues/3612) to track the addition of this improvement to the existing analyzer, plus another improvement I just thought of: If offset=0 and count=buffer.Length, then `AsMemory()` should have no parameters.

layomia commented on src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs:
```diff
@@ -674,11 +674,7 @@ private static async
-                        written,
-                        utf8BomLength - written,
-                        cancellationToken).ConfigureAwait(false);
+                    lastRead = await stream.ReadAsync(rented.AsMemory(written, utf8BomLength - written), cancellationToken).ConfigureAwait(false);
```
It doesn't - see [here](https://dev.azure.com/dnceng/public/_build/results?buildId=634534&view=logs&jobId=423ae147-26a3-5a4d-a4d3-4b3fc1dd598e&j=f3c44181-05e8-5372-473c-a59f5cdcfb6d&t=4e591485-5f7c-5f70-04d2-a4794c077c42) & [there](https://dev.azure.com/dnceng/public/_build/results?buildId=634534&view=logs&jobId=423ae147-26a3-5a4d-a4d3-4b3fc1dd598e&j=423ae147-26a3-5a4d-a4d3-4b3fc1dd598e&t=98a4dd5a-baea-5dc4-484b-ccbd315f07c8).

We should use an #if directive here, similar to https://github.com/dotnet/runtime/blob/03d428ef593762ed51779bec186d7f682b432c33/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L125-L131

carlossanlop replied:
I locally built `build.cmd clr+libs -rc release`, so I didn't see any netstandard related failures. I suspect this is why the CI failed in the PR. I'll apply @layomia 's suggestion.

layomia commented on src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs:
```diff
@@ -701,11 +697,7 @@ private static async
-                        written,
-                        rented.Length - written,
-                        cancellationToken).ConfigureAwait(false);
+                    lastRead = await stream.ReadAsync(rented.AsMemory(written, rented.Length - written), cancellationToken).ConfigureAwait(false);
```
Here as well.

carlossanlop replied:
Sure, no problem. BTW, this is a case that can be simplified as Levi pointed out above: `rented.AsMemory(written)`.

carlossanlop commented on src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextReaderImplAsync.cs:
```diff
@@ -1192,7 +1192,7 @@ private async Task<int> ReadDataAsync()
                     if (_ps.bytePos == _ps.bytesUsed && _ps.bytes.Length - _ps.bytesUsed > 0)
                     {
-                        int read = await _ps.stream.ReadAsync(_ps.bytes, _ps.bytesUsed, _ps.bytes.Length - _ps.bytesUsed).ConfigureAwait(false);
+                        int read = await _ps.stream.ReadAsync(_ps.bytes.AsMemory(_ps.bytesUsed, _ps.bytes.Length - _ps.bytesUsed)).ConfigureAwait(false);
```
Done and done.

GrabYourPitchforks reviewed (approved):
Thanks!

carlossanlop commented:
The unit tests for `XmlUtf8RawTextWriterAsync` are failing. I'm investigating.

buyaa-n commented on src/libraries/System.Private.Xml/src/System/Xml/Core/XmlEncodedRawTextWriterAsync.cs:
```diff
@@ -688,13 +688,13 @@ private async Task EncodeCharsAsync(int startOffset, int endOffset, bool writeAl
                 if (bufBytesUsed >= (bufBytes.Length - 16))
                 {
-                    await stream.WriteAsync(bufBytes, 0, bufBytesUsed).ConfigureAwait(false);
+                    await stream.WriteAsync(bufBytes.AsMemory(0, bufBytesUsed)).ConfigureAwait(false);
```
This is generated file, please update `XmlRawTextWriterGeneratorAsync.ttinclude` as well

carlossanlop replied:
Thanks, I didn't know about the generated files. I wonder if this is the reason why the CI is failing on these unit tests. I'll make the changes and then reply back with my findings.

buyaa-n replied:
I don't think the test failed from template, but the failures seems happening on the test i added recently, i will take a look what is wrong

buyaa-n replied:
Please add mock overloads to the test stream so that it wouldn't call  `MemoryStream.WriteAsync(ReadOnlyMemory<byte>, CancellationToken)`, the details sent by PM
```csharp
public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
{
    return default;
}
```

buyaa-n commented on src/libraries/System.Private.Xml/src/System/Xml/Core/XmlUtf8RawTextWriterAsync.cs:
```diff
@@ -552,7 +552,7 @@ protected virtual async Task FlushBufferAsync()
                     {
                         Debug.Assert(stream != null);
-                        await stream.WriteAsync(bufBytes, 1, bufPos - 1).ConfigureAwait(false);
+                        await stream.WriteAsync(bufBytes.AsMemory(1, bufPos - 1)).ConfigureAwait(false);
```
this one also has template

carlossanlop committed:
Apply CA1835 Roslyn fixes on .NET Libraries
- Prefer the Stream.ReadAsync and Stream.WriteAsync memory overloads.
- Unit test fix suggested by buyaa-n to adjust the usage of these overloads.

carlossanlop committed:
Additional missing unit test overload for ReadOnlyMemory

carlossanlop commented:
@buyaa-n the unit tests passed with your suggested additions. Thanks for your help. I'll merge this now.
